samza-kafka/src/main/scala/org/apache/samza/checkpoint/kafka/KafkaCheckpointManager.scala

```diff
@@ -19,258 +19,212 @@
-    if (partitionMetaData == null) {
-      throw new SamzaException("Cannot get partitionMetaData for system=%s, topic=%s" format(systemName, topic))
+    if (systemConsumer != null) {
+      systemConsumer.stop
     }
-
-    return partitionMetaData
+    info("CheckpointManager stopped.")
   }
 
   /**
-   * Reads an entry from the checkpoint log and invokes the provided lambda on it.
-   *
-   * @param handleEntry Code to handle an entry in the log once it's found
-   */
-  private def readLog(shouldHandleEntry: (KafkaCheckpointLogKey) => Boolean,
-                      handleEntry: (ByteBuffer, KafkaCheckpointLogKey) => Unit): Unit = {
-    info("Reading from checkpoint system:%s topic:%s" format(systemName, checkpointTopic))
+    * Returns the checkpoints from the log.
+    *
+    * <p> The underlying {@link SystemConsumer} is stateful and tracks its offsets. Hence, each invocation of this method
+    * will read the log from where it left off previously. This allows for multiple efficient calls to [[readLastCheckpoint()]]
+    */
+  private def readCheckpoints(): Map[TaskName, Checkpoint] = {
+    val checkpoints = mutable.Map[TaskName, Checkpoint]()
 
-    val ssp: SystemStreamPartition = new SystemStreamPartition(systemName, checkpointTopic, new Partition(0))
+    val iterator = new SystemStreamPartitionIterator(systemConsumer, checkpointSsp)
+    var numMessagesRead = 0
 
-    if (systemConsumer == null) {
-      val partitionMetadata = getSSPMetadata(checkpointTopic, new Partition(0))
-      val oldestOffset = partitionMetadata.getOldestOffset
-
-      systemConsumer = getSystemConsumer()
-      systemConsumer.register(ssp, oldestOffset)
-      systemConsumer.start()
-    }
-
-    val iterator =  new SystemStreamPartitionIterator(systemConsumer, ssp);
-    var msgCount = 0
     while (iterator.hasNext) {
-      val msg = iterator.next
-      msgCount += 1
-
-      val offset = msg.getOffset
-      val key = msg.getKey.asInstanceOf[Array[Byte]]
-      if (key == null) {
-        throw new KafkaUtilException(
-          "While reading checkpoint (currentOffset=%s) stream encountered message without key." format offset)
-      }
+      numMessagesRead += 1
+      val checkpointEnvelope: IncomingMessageEnvelope = iterator.next
 
-      val checkpointKey = KafkaCheckpointLogKey.fromBytes(key)
+      val keyBytes = checkpointEnvelope.getKey.asInstanceOf[Array[Byte]]
+      if (keyBytes == null) {
+        throw new SamzaException(s"Encountered a checkpoint message without key. Topic:$checkpointTopic " +
+          s"Offset:${checkpointEnvelope.getOffset}")
+      }
 
-      if (!shouldHandleEntry(checkpointKey)) {
-        info("Skipping checkpoint log entry at offset %s with key %s." format(offset, checkpointKey))
-      } else {
-        val checkpointPayload = ByteBuffer.wrap(msg.getMessage.asInstanceOf[Array[Byte]])
-        handleEntry(checkpointPayload, checkpointKey)
+      val checkpointKey = checkpointKeySerde.fromBytes(keyBytes)
```

Comment:
Catch and add context (e.g. offset, the string representation of bytes etc.)?

Reply:
Since, the string representation of bytes is now logged in the serde exception, I logged the offset here. Thanks for this feedback to improve debuggability!